File: leetcode/medium/919-cbt-inserter/q.py
"""
  @author Vortexx2 
  Problem 919 - CBT Inserter

  Runtime - 56 ms
  Memory Usage - 14.9 MB
"""
from collections import deque


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class CBTInserter:

  def __init__(self, root: TreeNode):
    self.dq = deque([])
    self.root = root
    q = deque([root])

    while q:
      node = q.popleft()

      if not node.left or not node.right:
        self.dq.append(node)

      if node.left:
        q.append(node.left)

      if node.right:
        q.append(node.right)

  def insert(self, v: int) -> int:
    node = self.dq[0]
    new_node = TreeNode(v)
    self.dq.append(new_node)

    if not node.left:
      node.left = new_node

    else:
      node.right = new_node
      self.dq.popleft()

    return node.val

  def get_root(self) -> TreeNode:
    return self.root

File: leetcode/medium/919-cbt-inserter/q.cpp
/**
 * @file q.cpp
 * @author Vortexx2
 * @brief Problem 919 CBT Inserter
 * @date 27-07-2021
 * 
 * Runtime - 12 ms
 * Memory Usage - 22.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <list>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class CBTInserter
{
public:
  TreeNode *root;
  queue<TreeNode *> q;
  CBTInserter(TreeNode *r) : root(r)
  {
    queue<TreeNode *> temp;
    temp.push(r);

    while (!temp.empty())
    {
      TreeNode *node = temp.front();
      temp.pop();

      if (!node->left || !node->right)
        q.push(node);

      if (node->left)
        temp.push(node->left);

      if (node->right)
        temp.push(node->right);
    }
  }

  int insert(int v)
  {
    TreeNode *newNode = new TreeNode(v);
    q.push(newNode);
    TreeNode *node = q.front();

    if (!node->left)
      node->left = newNode;

    else
    {
      node->right = newNode;
      q.pop();
    }

    return node->val;
  }

  TreeNode *get_root()
  {
    return root;
  }
};
